daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1316,6 +1316,173 @@ object SBuiltin {
     }
   }
 
+  // Unstable text primitives.
+
+  /** $text_to_upper :: Text -> Text */
+  final case object SBTextToUpper extends SBuiltin(1) {
+    def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
+      args.get(0) match {
+        case SText(t) =>
+          machine.ctrl = CtrlValue(SText(t.toUpperCase(util.Locale.ROOT)))
+            // TODO [FM]: replace with ASCII-specific function, or not
+        case x =>
+          throw SErrorCrash(s"type mismatch SBTextoUpper, expected Text got $x")
+      }
+    }
+  }
+
+  /** $text_to_lower :: Text -> Text */
+  final case object SBTextToLower extends SBuiltin(1) {
+    def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
+      args.get(0) match {
+        case SText(t) =>
+          machine.ctrl = CtrlValue(SText(t.toLowerCase(util.Locale.ROOT)))
+            // TODO [FM]: replace with ASCII-specific function, or not
+        case x =>
+          throw SErrorCrash(s"type mismatch SBTextToLower, expected Text got $x")
+      }
+    }
+  }
+
+  /** $text_slice :: Int -> Int -> Text -> Text */
+  final case object SBTextSlice extends SBuiltin(3) {
+    def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
+      args.get(0) match {
+        case SInt64(from) =>
+          args.get(1) match {
+            case SInt64(to) =>
+              args.get(2) match {
+                case SText(t) =>
+                  if (to <= 0 || from >= t.length.toLong || to <= from ) {
+                    machine.ctrl = CtrlValue(SText(""))
+                  } else {
+                    val rfrom = from.max(0).toInt
+                    val rto = to.min(t.length.toLong).toInt
+                      // NOTE [FM]: We use toInt only after ensuring the indices are
+                      // between 0 and t.length inclusive. Calling toInt prematurely
+                      // would mean dropping the high order bits indiscriminitely,
+                      // so for instance (0x100000000L).toInt == 0, resulting in an
+                      // empty string below even though `to` was larger than the
+                      // length.
+                    machine.ctrl = CtrlValue(SText(t.slice(rfrom, rto)))
```

Comment:
not working with Utf8

Reply:
That’s a fair point. Given that our goal is benchmarking we should try to roughly match the performance of the eventual operations even if we don’t support them in their full generality for now.

Reply:
Alternatively, we switch to using UTF16 strings like Java 🙃